sql: TestStatsWithLowTTL failed

sql.TestStatsWithLowTTL [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2963109&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2963109&tab=artifacts#/) on master @ [d302ba88430dd2438a5edcee15edf18286f81485](https://github.com/cockroachdb/cockroach/commits/d302ba88430dd2438a5edcee15edf18286f81485):


```
=== RUN   TestStatsWithLowTTL
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStatsWithLowTTL121354166
    test_log_scope.go:74: use -show-logs to present logs inline
E210508 18:21:22.808755 218020 jobs/adopt.go:260  [n1] 1  job 656761426508775425: adoption completed with error batch timestamp 1620498081.329064935,0 must be after replica GC threshold 1620498081.594562838,0
E210508 18:21:24.163191 218145 jobs/adopt.go:260  [n1] 2  job 656761431225761793: adoption completed with error batch timestamp 1620498082.963365067,0 must be after replica GC threshold 1620498083.047776328,0
    create_stats_test.go:142: pq: batch timestamp 1620498082.963365067,0 must be after replica GC threshold 1620498083.047776328,0
    create_stats_test.go:150: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStatsWithLowTTL121354166
--- FAIL: TestStatsWithLowTTL (5.04s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestStatsWithLowTTL PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries @RaduBerinde
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatsWithLowTTL.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestStatsWithLowTTL [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2992842&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2992842&tab=artifacts#/) on master @ [3ea71eb8ac28264e7b6e28404358966b19b50c35](https://github.com/cockroachdb/cockroach/commits/3ea71eb8ac28264e7b6e28404358966b19b50c35):


```
=== RUN   TestStatsWithLowTTL
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStatsWithLowTTL976399597
    test_log_scope.go:74: use -show-logs to present logs inline
E210518 12:52:08.056216 221389 jobs/adopt.go:260  [n1] 1  job 659527850281304065: adoption completed with error batch timestamp 1621342326.865111852,0 must be after replica GC threshold 1621342326.973265613,0
E210518 12:52:09.091974 221604 jobs/adopt.go:260  [n1] 2  job 659527853970194433: adoption completed with error batch timestamp 1621342327.991003571,0 must be after replica GC threshold 1621342328.073995353,0
    create_stats_test.go:142: pq: batch timestamp 1621342327.991003571,0 must be after replica GC threshold 1621342328.073995353,0
    create_stats_test.go:150: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStatsWithLowTTL976399597
--- FAIL: TestStatsWithLowTTL (4.51s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestStatsWithLowTTL PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries RaduBerinde
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStatsWithLowTTL.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/create_stats_test.go
+++ b/pkg/sql/create_stats_test.go
@@ -20,7 +20,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/security"
-	"github.com/cockroachdb/cockroach/pkg/sql/rowexec"
+	"github.com/cockroachdb/cockroach/pkg/sql/row"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
@@ -36,10 +36,11 @@ func TestStatsWithLowTTL(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	// The test requires a bunch of data to be inserted, which is much slower in
-	// race mode.
+	// The test depends on reasonable timings, so don't run under race.
 	skip.UnderRace(t)
 
+	// Set the KV batch size to 1 to avoid having to use a large number of rows.
+	defer row.TestingSetKVBatchSize(1)()
 	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{})
 	defer s.Stopper().Stop(context.Background())
 
@@ -50,7 +51,7 @@ func TestStatsWithLowTTL(t *testing.T) {
 		USE test;
 		CREATE TABLE t (k INT PRIMARY KEY, a INT, b INT);
 	`)
-	const numRows = 50000
+	const numRows = 5
 	r.Exec(t, `INSERT INTO t SELECT k, 2*k, 3*k FROM generate_series(0, $1) AS g(k)`, numRows-1)
 
 	pgURL, cleanupFunc := sqlutils.PGUrl(t,
@@ -61,8 +62,8 @@ func TestStatsWithLowTTL(t *testing.T) {
 	defer cleanupFunc()
 
 	// Start a goroutine that keeps updating rows in the table and issues
-	// GCRequests simulating a 1 second TTL. While this is running, reading at a
-	// timestamp older than 1 second will error out.
+	// GCRequests simulating a 2 second TTL. While this is running, reading at a
+	// timestamp older than 2 seconds will likely error out.
 	var goroutineErr error
 	var wg sync.WaitGroup
 	wg.Add(1)
@@ -96,9 +97,9 @@ func TestStatsWithLowTTL(t *testing.T) {
 				goroutineErr = err
 				return
 			}
-			// Force a table GC of values older than 1 second.
+			// Force a table GC of values older than 2 seconds.
 			if err := s.ForceTableGC(
-				context.Background(), "test", "t", s.Clock().Now().Add(-int64(1*time.Second), 0),
+				context.Background(), "test", "t", s.Clock().Now().Add(-int64(2*time.Second), 0),
 			); err != nil {
 				goroutineErr = err
 				return
@@ -107,10 +108,10 @@ func TestStatsWithLowTTL(t *testing.T) {
 		}
 	}()
 
-	// Sleep 500ms after every 10k scanned rows, to simulate a long-running
-	// operation.
-	rowexec.TestingSamplerSleep = 500 * time.Millisecond
-	defer func() { rowexec.TestingSamplerSleep = 0 }()
+	// Sleep 500ms after every scanned row (note that we set the KV batch size to
+	// 1 above), to simulate a long-running operation.
+	row.TestingInconsistentScanSleep = 500 * time.Millisecond
+	defer func() { row.TestingInconsistentScanSleep = 0 }()
 
 	// Sleep enough to ensure the table descriptor existed at AOST.
 	time.Sleep(100 * time.Millisecond)
@@ -134,8 +135,8 @@ func TestStatsWithLowTTL(t *testing.T) {
 		t.Log("expected CREATE STATISTICS to fail, trying again")
 	}
 
-	// Set up timestamp advance to keep timestamps no older than 0.3s.
-	r.Exec(t, `SET CLUSTER SETTING sql.stats.max_timestamp_age = '0.3s'`)
+	// Set up timestamp advance to keep timestamps no older than 1s.
+	r.Exec(t, `SET CLUSTER SETTING sql.stats.max_timestamp_age = '1s'`)
 
 	_, err := db.Exec(`CREATE STATISTICS foo FROM t AS OF SYSTEM TIME '-0.1s'`)
 	if err != nil {

--- a/pkg/sql/row/fetcher.go
+++ b/pkg/sql/row/fetcher.go
@@ -587,6 +587,13 @@ func (rf *Fetcher) StartScan(
 	return rf.StartScanFrom(ctx, &f)
 }
 
+// TestingInconsistentScanSleep introduces a sleep inside the fetcher after
+// every KV batch (for inconsistent scans, currently used only for table
+// statistics collection).
+// TODO(radu): consolidate with forceProductionKVBatchSize into a
+// FetcherTestingKnobs struct.
+var TestingInconsistentScanSleep time.Duration
+
 // StartInconsistentScan initializes and starts an inconsistent scan, where each
 // KV batch can be read at a different historical timestamp.
 //
@@ -645,6 +652,9 @@ func (rf *Fetcher) StartInconsistentScan(
 		res, err := txn.Send(ctx, ba)
 		if err != nil {
 			return nil, err.GoError()
+		}
+		if TestingInconsistentScanSleep != 0 {
+			time.Sleep(TestingInconsistentScanSleep)
 		}
 		return res, nil
 	}

--- a/pkg/sql/rowexec/sampler.go
+++ b/pkg/sql/rowexec/sampler.go
@@ -232,11 +232,6 @@ func (s *samplerProcessor) Run(ctx context.Context) {
 	s.MoveToDraining(nil /* err */)
 }
 
-// TestingSamplerSleep introduces a sleep inside the sampler, every
-// <samplerProgressInterval>. Used to simulate a heavily throttled
-// run for testing.
-var TestingSamplerSleep time.Duration
-
 func (s *samplerProcessor) mainLoop(ctx context.Context) (earlyExit bool, err error) {
 	rng, _ := randutil.NewPseudoRand()
 	var da rowenc.DatumAlloc
@@ -313,10 +308,6 @@ func (s *samplerProcessor) mainLoop(ctx context.Context) (earlyExit bool, err er
 					}
 				}
 				lastWakeupTime = timeutil.Now()
-			}
-
-			if TestingSamplerSleep != 0 {
-				time.Sleep(TestingSamplerSleep)
 			}
 		}
 

